Navigation block: adds in default and style variation

karmatosed opened the pull request:
Start of fixing #18552. The light style is a light grey to avoid it not being visible on light without a border. This is a very simple bubble drop down with a gentle arrow.

To do:

- [x] Dark style variation
- [x] Remove text-decoration (bug)

Commit:
Adds in default style variation

Start of fixing #18552

Commit:
Dark styling

karmatosed commented:
This is now ready to review as I will work on iterations later. It would be great to get in along with experimental flag being removed.

draganescu commented:
The styles appeared but switching showed no difference (using the TwentyTwenty theme) although I see there is some styling present for the dark style.

karmatosed commented:
@draganescu can you code review as I think this can be easier worked out with https://github.com/WordPress/gutenberg/pull/18466 - which doesn't work on this branch. Or if there's a way to get that picked up on branch I think we can get this working.

karmatosed commented:
This is what by default is in Twenty Nineteen:

![image](https://user-images.githubusercontent.com/253067/69173294-653e4100-0af7-11ea-9cb7-d11ae1615921.png)

jasmussen commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -61,9 +61,8 @@
 		& > li > ul {
 			margin: 0;
 			position: absolute;
-			background: #fff;
```
I think it is _correct_ to remove this background, because it is not "light or dark theme aware". 

But it does mean that a theme like TwentyTwenty now looks like this:

<img width="551" alt="Screenshot 2019-11-20 at 08 44 05" src="https://user-images.githubusercontent.com/1204802/69218913-0669dd80-0b72-11ea-845c-85a4bb50b706.png">

Compare that with this before:

<img width="499" alt="Screenshot 2019-11-20 at 08 45 08" src="https://user-images.githubusercontent.com/1204802/69218967-226d7f00-0b72-11ea-8dc1-6a354c75ecb2.png">

This is a tricky one, because the white background assumes a link color that has contrast with that which might not be the case. On the other hand just floating there is also a bit off. Any ideas?

jasmussen replied:
Note that I tested TwentyTwenty because it does _not_ opt in to "opinionated styles", and your PR here correctly relegates most of those styles to be actually opinionated.

jasmussen commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -61,9 +61,8 @@
-			background: #fff;
 			left: 0;
-			top: 100%;
+			top: 80%;
```
What does this do?

karmatosed replied:
This moves sub-nav closer to the navigation.

jasmussen commented on packages/block-library/src/navigation/theme.scss:
```diff
@@ -3,4 +3,45 @@
 		list-style: none;
 	}
+
+	ul > li > ul {
```
I'm 90% sure that this is necessary because menus can nest deeply, so you have to be specific in targetting. But are there any CSS classes you can use to target this instead? Like, will `.wp-block-navigation-item > ul` work?

The fewer items in selectors, the better!

jasmussen commented on packages/block-library/src/navigation/theme.scss:
```diff
@@ -3,4 +3,45 @@
 	}
+
+	ul > li > ul {
+		background: #e3e3e3;
```
I think this is good. But if the goal is to make it stand out from the background, it won't work if the background is also gray :)

So we can either embrace that it might blend in, in which case I'd personally love a sliiiightly lighter gray. 

Or we can do something like add a shadow, gasp:

<img width="543" alt="Screenshot 2019-11-20 at 08 52 08" src="https://user-images.githubusercontent.com/1204802/69219473-349bed00-0b73-11ea-880e-3d1ef44bd7b0.png">

Up to you!

jasmussen commented on packages/block-library/src/navigation/theme.scss:
```diff
@@ -3,4 +3,45 @@
+
+	ul > li > ul {
+		background: #e3e3e3;
+		border-radius: 4px;
```
I wish we had some sort of system for cataloguing these opinionated styles. Like, a consistent coherent visual identity for all WordPress opinionated styles that blocks can opt into. Right now they are kind of designed in isolation, but it'd be nice to have a system similar to the variables sheet we have for UI.

Things like color variables, border radii, paddings, etc. Not for this PR, but CC: @ItsJonQ as I think you've been vaguely exploring things like these?

ItsJonQ replied:
@jasmussen Ah yes! I outlined my idea here:
https://github.com/WordPress/gutenberg/issues/9534#issuecomment-555173108

jasmussen reviewed (commented):
I left some comments, but this seems like the right approach: 

- It adds the designs as style variations you can choose from.
- It embraces the theme.scss file because they are opinionated styles.
- Code seems good, works well.

There's still the accessibility challenge of tabbing through the menu items, but that's separate. 

There's an issue with the variation preview itself:

<img width="302" alt="Screenshot 2019-11-20 at 08 56 15" src="https://user-images.githubusercontent.com/1204802/69219750-cc99d680-0b73-11ea-8074-c99d9cb646a6.png">

I don't know what we can do here. I know we can provide some default example content here, Image block does that, but the change is only visible when the mouse hovers the menu item, which doesn't show up. 

@gziolo can we show not the nav block but JUST the submenu here? I.e. custom markup for the preview?

FINALLY I think there's an open question whether this PR should make THREE styles.

- Default: which is just the white background I commented on, and nothing else.
- Light: the gray bubble
- Dark: the dark bubble

This gives a little choice?

Something to think about.

Commit:
Merge branch 'try/nav-styles' of https://github.com/WordPress/gutenberg into try/nav-styles

Brings a default to this but still has issues with mobile styling and also unsure what belongs in theme or what style to allow for variations.

@jasmussen love your feedback on this approach. I can only think of this to have a default for Twenty Twenty being white with box-shadow.

What seems to still be issue is the class being passed and it changing.

gziolo commented:
> @gziolo can we show not the nav block but JUST the submenu here? I.e. custom markup for the preview?

I didn't follow the progress on Navigation block so I don't think I can help without taking a deep dive. I'm also attending WordCamp this week so I'd suggest asking someone else for help.

jasmussen commented:
Took this for a spin again, and I think I have an idea for a good path forward.

However for some reason I can't get the styles to work anymore, not sure why. The code looks right, but for whatever, `is-style-default` and `is-style-dark` CSS classes do not get added to the navigation menu either in TwentyNineteen nor TwentyTwenty, which they should, so the styles do not show up. It even looks like `.wp-block-navigation-menu` is gone from the saved preview. I don't know if that disappeared in master while you were working on this, not sure what is going on there. So this is a place where we might need expert advice from someone else.

However, once that is fixed up, I think you should do these things to get this shipping:

1. Let's have your two styles be _in addition_ to the default style. So, have `default` (what's shipping today) `light`, and `dark`.
2. Move all the styles from theme.scss to style.scss. By adding the two styles as _additional_ styles that are non-default, it is okay to ship these without theme opt-in, because it becomes _user-opt in_. I would also argue that because they are style variations, they have to be in style.scss, otherwise TwentyTwenty users who explicitly choose those style variations don't see any change at all. 
3. We need to create a new ticket (or find an existing one?) for showing _custom content_ in the Preview windows. Right now you can only _customize the block you're previewing_, you can't add custom _content_. This is POSSIBLY intentional, but it's also the only way I can think of, to actually show the on-hover bubble in the preview.

Does this make sense? What do you think?

karmatosed commented:
> Does this make sense? What do you think?

This sounds great. Thanks so much.

@youknowriad are you able to or someone else help fix through the issue here?

> However for some reason I can't get the styles to work anymore, not sure why. The code looks right, but for whatever, is-style-default and is-style-dark CSS classes do not get added to the navigation menu either in TwentyNineteen nor TwentyTwenty, which they should, so the styles do not show up. It even looks like .wp-block-navigation-menu is gone from the saved preview. I don't know if that disappeared in master while you were working on this, not sure what is going on there. So this is a place where we might need expert advice from someone else

mtias commented:
I think the default style should be called "Light".

karmatosed commented:
> I think the default style should be called "Light".

Sounds great, once through this class not showing issue will do that.

youknowriad committed:
Fix frontend className generation

youknowriad commented:
Pushed a fix for the frontend className generation.

youknowriad commented:
I also rebased to fix some conflicts.

karmatosed commented:
Thank you so much @youknowriad, you are ace! I will pick it up and do the other little things.

Commit:
Updates to PR

- Includes a fix for padding
- Makes default light
- Moves all to styles.scss to avoid Twenty Twenty issues.

Commit:
Add in border radius on default

karmatosed commented:
Screenshots of latest iterations:

![image](https://user-images.githubusercontent.com/253067/69552059-b46ff000-0f95-11ea-8165-bc30e83d295c.png)

![image](https://user-images.githubusercontent.com/253067/69552082-bf2a8500-0f95-11ea-8bec-1cc05fc7e083.png)

shaunandrews commented:
<img width="177" alt="image" src="https://user-images.githubusercontent.com/191598/69555449-37328400-0f71-11ea-84fe-d2ba10dc46c1.png">

Looks like the pointer arrow is a little borked.

<img width="134" alt="image" src="https://user-images.githubusercontent.com/191598/69555484-4580a000-0f71-11ea-8b95-c903f2e02106.png">

The space between items in the dropdown feels _way_ too big to my eyes. With the current spacing, a list of 5-6 links is going to make the dropdown unwieldy in height.

Commit:
Fixes issue on chrome

Chrome in screenshot was showing bottom border out on arrow. Fixed now. Props @mtias for noticing 🙌🏻

karmatosed commented:
I have resolved the issue with the arrow in the latest PR.

![image](https://user-images.githubusercontent.com/253067/69556332-73c7a500-0f9c-11ea-9d4b-268c60cd2a37.png)

karmatosed commented:
Uploading a PR with slightly some spacing adjustments to result in:

![image](https://user-images.githubusercontent.com/253067/69556813-5ba45580-0f9d-11ea-8c2c-a64d563203d4.png)

Commit:
Adapts spacing a little

Reduces on lines themselves and adds wrapper spacing within navigation when sub panel.

youknowriad reviewed (approved):
LGTM code wise.
